### PR TITLE
[EE] Unify all classes of CSharpExpressionCompiler

### PR DESCRIPTION
**Pull request by ivanbasov**
fixes https://github.com/dotnet/roslyn/issues/22620

### CODE REVIEW

**Comment by ivanbasov** [comment]
tagging @r-ramesh for review

**Comment by ivanbasov** [comment]
@r-ramesh is added to the review. #Closed

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs [hunk 1]
```diff
@@ -4,16 +4,30 @@
...
+using Microsoft.VisualStudio.Debugger.CallStack;
 using Microsoft.VisualStudio.Debugger.Clr;
+using Microsoft.VisualStudio.Debugger.ComponentInterfaces;
 using Microsoft.VisualStudio.Debugger.Evaluation;
 
 namespace Microsoft.CodeAnalysis.CSharp.ExpressionEvaluator
 {
     [DkmReportNonFatalWatsonException(ExcludeExceptionType = typeof(NotImplementedException)), DkmContinueCorruptingException]
-    internal sealed class CSharpExpressionCompiler : ExpressionCompiler
+    internal sealed class CSharpExpressionCompiler : ExpressionCompiler, IDkmLanguageFrameDecoder, IDkmLanguageInstructionDecoder
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs]: The same change will be needed for `VisualBasicExpressionCompiler`.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs [hunk 2]
```diff
@@ -4,16 +4,30 @@
...
     {
         private static readonly DkmCompilerId s_compilerId = new DkmCompilerId(DkmVendorId.Microsoft, DkmLanguageId.CSharp);
 
+        // Need to support IDkmLanguageFrameDecoder and IDkmLanguageInstructionDecoder
+        // See https://github.com/dotnet/roslyn/issues/22620
+        private CSharpFrameDecoder _frameDecoder;
+        private CSharpLanguageInstructionDecoder _languageInstructionDecoder;
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs]: `readonly`

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs (hunk 1 above)
Comment (cston) [reply src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs]: Consider moving the interface implementations and the `FrameDecoder` and `LanguageInstructionDecoder` fields to the base class.

Comment (ivanbasov) [reply src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs]: Thank you!

---
In reply to: [144460725](https://github.com/dotnet/roslyn/pull/22681#discussion_r144460725) [](ancestors = 144460725)

Comment (ivanbasov) [reply src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs]: Thank you!

---
In reply to: [144460858](https://github.com/dotnet/roslyn/pull/22681#discussion_r144460858) [](ancestors = 144460858)

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs (hunk 2 above)
Comment (ivanbasov) [reply src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpExpressionCompiler.cs]: Thank you!

---
In reply to: [144460747](https://github.com/dotnet/roslyn/pull/22681#discussion_r144460747) [](ancestors = 144460747)

**Comment by ivanbasov** [comment]
@r-ramesh, I have various issues with verification the fix locally. Could you please review and if it looks good help with verification? Thank you!

**Comment by r-ramesh** [comment]
Looks good to me. @plnelson might have some feedback.

**Comment by plnelson** [comment]
LGTM

**Comment by ivanbasov** [comment]
A similar PR is merged for 15.6. I would prefer not to merge this into 15.5 to give more time for verification.

**Pull request closed without merging**